src/main/scala/org/ergoplatform/api/WalletApiRoute.scala

```diff
@@ -38,16 +42,38 @@ case class WalletApiRoute(readersHolder: ActorRef, nodeViewActorRef: ActorRef, e
       generateAssetIssueTransactionR ~
       sendTransactionR ~
       sendPaymentTransactionR ~
-      sendAssetIssueTransactionR
+      sendAssetIssueTransactionR ~
+      p2shAddressR ~
+      p2sAddressR
   }
 
+  private val loadMaxKeys: Int = 100
+
   private val fee: Directive1[Option[Long]] = entity(as[Json]).flatMap { p =>
     Try(p.hcursor.downField("fee").as[Long]) match {
       case Success(Right(value)) => provide(Some(value))
       case _ => provide(None)
     }
   }
 
+  private val source: Directive1[String] = entity(as[Json]).flatMap { p =>
+    p.hcursor.downField("source").as[String]
+      .fold(_ => reject, s => provide(s))
+  }
+
+  private def addressResponse(address: ErgoAddress): Json = Json.obj("address" -> addressJsonEncoder(address))
+
+  private def compileSource(source: String, publicKeys: Seq[ProveDlog]): Try[Value[SBoolean.type]] = {
```

Comment:
@oskin1 why not allow env: Map[String, Any] in the place of publicKeys? Also what about ProveDHT?

Reply:
@ergomorphic, the idea is only to load available in wallet public keys into env, so why do we really need `env: Map[String, Any]` here?

Reply:
@ergomorphic, ok, I've fixed that.